docs/models/advanced/imodel.md

```diff
@@ -0,0 +1,159 @@
+
+If an `IModel` also implements certain subinterfaces of `IModel`, it can also process some extra information. External entities using the model can do this processing through the static helpers in `ModelProcessingHelper`.
+
+At this high level, a model has no concept of items, blocks, or anything of that sort; it purely represents a shape.
+
+!!! important
+    `IModel` is immutable. Methods such as `IModelCustomData::process` that alter the model should never modify the `IModel`, as they should construct new `IModel`s instead.
+
+### `getDependencies`
+
+This is a collection of the `ResourceLocation`s of all the models this model depends on. These models are guaranteed to be loaded before this one is baked. For example, a model deserialized from a blockstate JSON will depend on the models defined within. Only models that are directly mapped to a block/item are loaded normally; to ensure loading of other models, they must be declared as dependencies of another. Cyclic dependencies will cause a `LoaderException` to be thrown.
+
+### `getTextures`
+
+This is a collection of the `ResourceLocation`s of all the textures this model depends on. These textures are guaranteed to be loaded before this model is baked. For example, a vanilla JSON model depends on all the textures defined within.
+
+### `bake`
+
+This is the main method of `IModel`. It takes an `IModelState`, a `VertexFormat`, and a function `ResourceLocation` → `TextureAtlasSprite`, to return an `IBakedModel`. `IBakedModel` is lower-level than `IModel`, and it is what interacts with blocks and items. The function RL → TAS is used to get textures from RLs (i.e. the RLs of textures are passed to this function and the returned TAS contains the texture).
+
+`IAnimatedModel`
+----------------
+
+`IAnimatedModel` is part of the currently WIP animation API. Right now it does nothing.
+
+`IModelCustomData`
+------------------
+
+`IModelCustomData` allows a model to process extra data from external sources. The Forge blockstate variant format provides a way to define this data in the resource pack. Within the Forge blockstate format, the property that is used to pass this data is called `custom`. First, an example:
+
+```json
+{
+  "forge_marker": 1,
+  "defaults": {
+    "custom": {
+      "__comment": "These can be any JSON primitives with any names, but models should only use what they understand.",
+      "meaningOfLife": 42,
+      "showQuestion": false
+    },
+    "__comment": "This model will receive the data in custom iff it is an `ICustomModelData`.",
+    "model": "examplemod:life_meaning"
+  },
+  "variants": {
+    "dying": {
+      "true": {
+        "__comment": "Custom data is inherited. Therefore, here `meaningOfLife` is inherited but `showQuestion` is overriden. The model itself remains inherited.",
+        "custom": {
+          "showQuestion": true
+        }
+      },
+      "false": {}
+    },
+    "book": {
```

Comment:
I wanted to show it worked in submodels too, and didn't think of a way to cram it into `"false"`, but if you think its unnecessary I'll get rid of it.